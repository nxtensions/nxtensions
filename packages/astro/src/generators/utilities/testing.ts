import { Tree, readJsonFile, updateJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace as _createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { PackageJson } from '../../utilities/package-json';

export function createTreeWithEmptyWorkspace(nxVersion?: string): Tree {
  const tree = _createTreeWithEmptyWorkspace();
  if (!nxVersion) {
    const { devDependencies } = readJsonFile<PackageJson>('package.json');
    nxVersion = devDependencies['nx'];
  }
  updateJson(tree, 'package.json', (json) => {
    json.devDependencies['nx'] = nxVersion;
    return json;
  });

  return tree;
}
