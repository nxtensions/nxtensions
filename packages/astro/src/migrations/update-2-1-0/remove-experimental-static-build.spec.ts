import {
  addProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import removeExperimentalStaticBuild from './remove-experimental-static-build';

describe('remove-experimental-static-build migration', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  describe('project configuration', () => {
    test('should do nothing when there are no executors meant to be migrated', async () => {
      const targets = {
        build: {
          executor: '@my-org/awesome-package:build',
          options: { experimentalStaticBuild: true },
        },
        dev: {
          executor: '@my-org/awesome-package:dev',
          options: { experimentalStaticBuild: true },
        },
      };
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets,
      });

      await removeExperimentalStaticBuild(tree);

      const project = readProjectConfiguration(tree, 'app1');
      expect(project.targets).toStrictEqual(targets);
    });

    test('should do nothing when the experimentalStaticBuild option is not set', async () => {
      const targets = {
        build: {
          executor: '@nxtensions/astro:build',
          options: { drafts: true },
        },
        dev: {
          executor: '@nxtensions/astro:dev',
          options: { drafts: true },
        },
      };
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets,
      });

      await removeExperimentalStaticBuild(tree);

      const project = readProjectConfiguration(tree, 'app1');
      expect(project.targets).toStrictEqual(targets);
    });

    test('should delete the experimentalStaticBuild option when set to true', async () => {
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets: {
          build: {
            executor: '@nxtensions/astro:build',
            options: { experimentalStaticBuild: true },
          },
          dev: {
            executor: '@nxtensions/astro:dev',
            options: { experimentalStaticBuild: true },
          },
        },
      });

      await removeExperimentalStaticBuild(tree);

      const project = readProjectConfiguration(tree, 'app1');
      expect(project.targets.build.options).toStrictEqual({});
      expect(project.targets.dev.options).toStrictEqual({});
    });

    test('should set legacyBuild to true when the experimentalStaticBuild option is set to false', async () => {
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets: {
          build: {
            executor: '@nxtensions/astro:build',
            options: { experimentalStaticBuild: false },
          },
          dev: {
            executor: '@nxtensions/astro:dev',
            options: { experimentalStaticBuild: false },
          },
        },
      });

      await removeExperimentalStaticBuild(tree);

      const project = readProjectConfiguration(tree, 'app1');
      expect(project.targets.build.options).toStrictEqual({
        legacyBuild: true,
      });
      expect(project.targets.dev.options).toStrictEqual({ legacyBuild: true });
    });

    test('should also update configurations', async () => {
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets: {
          build: {
            executor: '@nxtensions/astro:build',
            options: { experimentalStaticBuild: false },
            configurations: {
              production: { experimentalStaticBuild: true },
            },
          },
          dev: {
            executor: '@nxtensions/astro:dev',
            options: {},
            configurations: {
              development: { silent: true, experimentalStaticBuild: false },
              production: { experimentalStaticBuild: true },
            },
            defaultConfiguration: 'development',
          },
        },
      });

      await removeExperimentalStaticBuild(tree);

      const project = readProjectConfiguration(tree, 'app1');
      expect(project.targets.build.options).toStrictEqual({
        legacyBuild: true,
      });
      expect(project.targets.build.configurations).toStrictEqual({
        production: {},
      });
      expect(project.targets.dev.options).toStrictEqual({});
      expect(project.targets.dev.configurations).toStrictEqual({
        development: { silent: true, legacyBuild: true },
        production: {},
      });
    });
  });

  describe('astro configuration', () => {
    beforeEach(() => {
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        projectType: 'application',
        targets: { build: { executor: '@nxtensions/astro:build' } },
      });
    });

    test('should do nothing when experimentalStaticBuild is not set', async () => {
      tree.write(
        'apps/app1/astro.config.mjs',
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            drafts: true,
          },
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(
        tree.read('apps/app1/astro.config.mjs', 'utf-8')
      ).toMatchSnapshot();
    });

    test('should do nothing when experimentalStaticBuild is contained in a comment', async () => {
      tree.write(
        'apps/app1/astro.config.mjs',
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            // experimentalStaticBuild: true,
            drafts: true,
          },
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(
        tree.read('apps/app1/astro.config.mjs', 'utf-8')
      ).toMatchSnapshot();
    });

    test('should do nothing when experimentalStaticBuild is not part of buildOptions (invalid config)', async () => {
      tree.write(
        'apps/app1/astro.config.mjs',
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            drafts: true,
          },
          experimentalStaticBuild: true,
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(
        tree.read('apps/app1/astro.config.mjs', 'utf-8')
      ).toMatchSnapshot();
    });

    test('should delete the experimentalStaticBuild option when set to true', async () => {
      tree.write(
        'apps/app1/astro.config.mjs',
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            experimentalStaticBuild: true,
            drafts: true,
          },
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(
        tree.read('apps/app1/astro.config.mjs', 'utf-8')
      ).toMatchSnapshot();
    });

    test('should set legacyBuild to true when the experimentalStaticBuild option is set to false', async () => {
      tree.write(
        'apps/app1/astro.config.mjs',
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            experimentalStaticBuild: false,
            drafts: true,
          },
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(
        tree.read('apps/app1/astro.config.mjs', 'utf-8')
      ).toMatchSnapshot();
    });

    test.each([
      'astro.config.mjs',
      'astro.config.js',
      'astro.config.cjs',
      'astro.config.ts',
    ])('should handle "%s"', async (configFile) => {
      tree.write(
        `apps/app1/${configFile}`,
        `export default /** @type {import('astro').AstroUserConfig} */ ({
          buildOptions: {
            experimentalStaticBuild: true,
            drafts: true,
          },
        });`
      );

      await removeExperimentalStaticBuild(tree);

      expect(tree.read(`apps/app1/${configFile}`, 'utf-8')).toMatchSnapshot();
    });
  });
});
