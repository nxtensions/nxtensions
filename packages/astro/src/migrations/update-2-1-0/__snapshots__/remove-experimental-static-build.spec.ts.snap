// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove-experimental-static-build migration astro configuration should delete the experimentalStaticBuild option when set to true 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should do nothing when experimentalStaticBuild is contained in a comment 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    // experimentalStaticBuild: true,
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should do nothing when experimentalStaticBuild is not part of buildOptions (invalid config) 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
  experimentalStaticBuild: true,
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should do nothing when experimentalStaticBuild is not set 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should handle "astro.config.cjs" 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should handle "astro.config.js" 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should handle "astro.config.mjs" 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    drafts: true,
  },
});
"
`;

exports[`remove-experimental-static-build migration astro configuration should handle "astro.config.ts" 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ {
  buildOptions: {
    drafts: true,
  },
};
"
`;

exports[`remove-experimental-static-build migration astro configuration should set legacyBuild to true when the experimentalStaticBuild option is set to false 1`] = `
"export default /** @type {import('astro').AstroUserConfig} */ ({
  buildOptions: {
    legacyBuild: true,
    drafts: true,
  },
});
"
`;
